tests/sycl/buffer.cpp

```diff
@@ -278,4 +278,161 @@ BOOST_AUTO_TEST_CASE(buffer_container_constructor_no_def_constr) {
     BOOST_CHECK(data2[i].val == testVal.val);
 }
 
+BOOST_AUTO_TEST_CASE(buffer_shared_ptr) {
+  namespace s = cl::sycl;
+  s::queue q{};
+
+  const std::size_t size = 1024;
+  int testVal = 42;
+
+  // Constructor that takes non-empty shared_ptr
+  {
+    std::shared_ptr<int> hostptr{new int[size], std::default_delete<int[]>{}};
+
+    {
+      s::buffer<int> buf{hostptr, size};
+
+      // buffer must copy the shared_ptr, so its ref count should be two now
+      BOOST_CHECK(hostptr.use_count() == 2);
+
+      q.submit([&](auto &cgh) {
+        auto acc = buf.get_access<s::access::mode::write>(cgh);
+        cgh.parallel_for(size, [=](auto idx) { acc[idx] = testVal; });
+      });
+    }
+
+    std::vector<int> hostdata(hostptr.get(), hostptr.get() + size);
+    for (auto val : hostdata)
+      BOOST_CHECK(val == testVal);
+  }
+
+  // Constructor that takes empty shared_ptr
+  {
+    std::shared_ptr<int> hostptr;
+    s::buffer<int> buf{hostptr, size};
+  }
+
+  // Constructor that takes empty shared_ptr, but set final data pointer
+  {
+    std::vector<int> hostdata(size, -1);
+    std::shared_ptr<int> hostptr;
+
+    {
+      s::buffer<int> buf{hostptr, size};
+      buf.set_final_data(hostdata.data());
+
+      q.submit([&](auto &cgh) {
+        auto acc = buf.get_access<s::access::mode::write>(cgh);
+        cgh.parallel_for(size, [=](auto idx) { acc[idx] = testVal; });
+      });
+    }
+
+    for (auto val : hostdata)
+      BOOST_CHECK(val == testVal);
+  }
+
+  // Constructor that takes a unique ptr
+  {
+    std::unique_ptr<int, std::default_delete<int[]>> hostptr{
+        new int[size], std::default_delete<int[]>{}};
+
+    s::buffer<int> buf{std::move(hostptr), size};
```

Comment:
@nilsfriess which overload is supposed to actually handle this?
Some MacOS build with old-ish libc++ isn't able to resolve to a specific constructor....

Reply:
I would have thought it chooses one of the constructors that takes a `std::shared_ptr<T>` as first argument (in this case the one without the `AllocatorT` argument).

Reply:
https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1609/commits/4041e9cd8763cabdd60e3692fd3cbd0b9a364bb2 helps for me.. Seems libc++ doesn't have this participate in overload resolution.

Reply:
But this is already tested (in the test starting in line 288), I added this particular test because the SYCL spec states the following in Sec. 4.7.2.3.:

> Note that since there is an implicit conversion of a std::unique_ptr to a std::shared_ptr, a std::unique_ptr can also be used to pass the ownership to the SYCL runtime. 

Maybe you could try something like this

```cpp
    const auto make_test_ptr = [size]() {
        return std::unique_ptr<int, std::default_delete<int[]>>{
            new int[size], std::default_delete<int[]>{}};
    };

    s::buffer<int> buf{make_test_ptr(), size};
```